src/cryptsetup/cryptsetup-tokens/luks2-tpm2.c

```diff
@@ -58,7 +58,7 @@ int acquire_luks2_key(
         if ((flags & TPM2_FLAGS_USE_PIN) && salt && !pin)
                 return -ENOANO;
 
-        if (pin) {
+        if (pin && salt && salt_size > 0) {
```

Comment:
Hmm. It seems the current logic require some thing like the following:
```c
assert(!!salt == salt_size > 0);
```

IIUC, the salt is provided by `tpm2_parse_luks2_json()`. Maybe, it is better that the function returns `NULL` for `salt` when `salt_size == 0`, and add the assertion in the above at the beginning of functions which take `salt` and `salt_size`.

Reply:
It does, checking both is just extra caution, as it doesn't cost anything

Reply:
> It does,

What does what? At least, `tpm2_parse_luks2_json()` provides non-NULL salt with zero length, IIUC.

Reply:
If there is a salt in the json, then the lenght will be non-zero, that's fine. The issue is that salt_size is _always_ returned, because it's initialized to zero, but we are not checking here and thus we go kaboom in case it is _not_ in the json token, like the case of a token enrolled < 253 and unlocked with >= 253

Reply:
I cannot follow. Maybe I am tired today. I will consider this later again if no one review it.